### PR TITLE
Support <classpath> in doc in the same way as for compile

### PR DESCRIPTION
**Pull request by tombentley**
This PR adds support for `<classpath>` to the `doc` and `doc-jar` goals. This can be useful in some niche cases where it is needed to omit some dependencies from the `<dependencies>` section of a POM, but still need them on the classpath at compile time.

### CODE REVIEW

**Comment by slandelle** [comment]
Does maven-javadoc-plugin have something similar?

**Comment by tombentley** [comment]
@slandelle that's a good point which I've now implemented.

It does introduce an inconsistency with the compile goal though, where this has to be done using `<classpath>`, which means the dependency won't be resolved for you, which is kind of a shame. I don't know what your objectives are with respect to conforming to maven plugin idioms, but it would be neat for me personally if the compile goal also supported `<additionalDependencies>`. If that's acceptable I'd be happy to add it in this PR or another.

**Comment by slandelle** [comment]
> I don't know what your objectives are with respect to conforming to maven plugin idioms

IMHO, idioms and consistency across an ecosystem are paramount. Things are just not this way as this plugin has a long history, with original developers no longer active :)

> but it would be neat for me personally if the compile goal also supported <additionalDependencies>. 

I completely agree. Let's deprecate `classpath`.

> If that's acceptable I'd be happy to add it in this PR or another.

It would be awesome

File: src/main/java/scala_maven/AdditionalDependency.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package scala_maven;
+
+import org.apache.maven.model.Dependency;
+
+public class AdditionalDependency extends Dependency {
```
Comment (slandelle) [inline src/main/java/scala_maven/AdditionalDependency.java:21]: Why do you need this?

File: src/main/java/scala_maven/ScalaDocMojo.java [hunk 2]
```diff
@@ -173,6 +173,9 @@
     @Parameter(defaultValue = "${project.build.sourceDirectory}/../scala")
     private File sourceDir;
 
+    @Parameter(property = "additionalDependencies")
+    private AdditionalDependency[] additionalDependencies;
```
Comment (slandelle) [inline src/main/java/scala_maven/ScalaDocMojo.java:177]: Why not `Dependency[]`?

**Comment by davidB** [comment]
classpath is a legacy from age where scala's libs (compiler,....) were not published into a maven repository ;-) (this plugin and the first maven repository scala-tools.org born together in 2008 iirc)

**Comment by slandelle** [comment]
Thanks for the explanation @davidB 
I meant no offense, just explaining that this plugin has a very long history :)
Are you fine with the deprecation?

**Comment by tombentley** [comment]
I've added support for `additionalDependencies` to the compile goal and deprecated `classpath` there, assuming @davidB is OK with that.

File: src/main/java/scala_maven/ScalaCompileMojo.java [hunk 3]
```diff
@@ -39,11 +39,21 @@
...
+    @Deprecated
     private Classpath classpath;
 
+    /**
+     * Additional dependencies to be added to the classpath. This can be useful in
+     * situations where a dependency is needed at compile time, but should not be
+     * treated as a dependency in the published POM.
+     */
+    @Parameter(property = "additionalDependencies")
+    private Dependency[] additionalDependencies;
```
Comment (slandelle) [inline src/main/java/scala_maven/ScalaCompileMojo.java:55]: Should we move this in `ScalaMojoSupport` parent and avoid code duplication?

File: src/main/java/scala_maven/ScalaCompileMojo.java [hunk 4]
```diff
@@ -59,8 +69,13 @@
     protected List<String> getClasspathElements() throws Exception {
         final List<String> back = project.getCompileClasspathElements();
         back.remove(project.getBuild().getOutputDirectory());
-        // back.add(getOutputDir().getAbsolutePath());
+        if (additionalDependencies != null) {
+            for (Dependency dependency : additionalDependencies) {
+                addToClasspath(factory.createDependencyArtifact(dependency), back, false);
+            }
+        }
```
Comment (slandelle) [inline src/main/java/scala_maven/ScalaCompileMojo.java:76]: Should we move this in `ScalaMojoSupport` parent and avoid code duplication?

File: src/main/java/scala_maven/ScalaDocMojo.java [hunk 5]
```diff
@@ -173,6 +174,14 @@
     @Parameter(defaultValue = "${project.build.sourceDirectory}/../scala")
     private File sourceDir;
 
+    /**
+     * Additional dependencies to be added to the classpath. This can be useful in
+     * situations where a dependency is needed at compile time, but should not be
+     * treated as a dependency in the published POM.
+     */
+    @Parameter(property = "additionalDependencies")
+    private Dependency[] additionalDependencies;
```
Comment (slandelle) [inline src/main/java/scala_maven/ScalaDocMojo.java:183]: Should we move this in `ScalaMojoSupport` parent and avoid code duplication?

File: src/main/java/scala_maven/ScalaDocMojo.java [hunk 6]
```diff
@@ -295,6 +304,13 @@ protected JavaMainCaller getScalaCommand() throws Exception {
         List<String> paths = new ArrayList<>(project.getCompileClasspathElements());
         paths.remove(project.getBuild().getOutputDirectory()); // remove output to avoid "error for" : error: XXX is
                                                                // already defined as package XXX ... object XXX {
+
+        if (additionalDependencies != null) {
+            for (Dependency dependency : additionalDependencies) {
+                addToClasspath(factory.createDependencyArtifact(dependency), paths, false);
+            }
+        }
```
Comment (slandelle) [inline src/main/java/scala_maven/ScalaDocMojo.java:312]: Should we move this in `ScalaMojoSupport` parent and avoid code duplication?

**Comment by davidB** [comment]
no offense, just the opportunity to share part of my knowledge ;-) . I'm fine with every clean, improvement,...

File: src/main/java/scala_maven/ScalaMojoSupport.java [hunk 7]
```diff
@@ -514,7 +515,7 @@ private void addToClasspath(String groupId, String artifactId, String version, S
         addToClasspath(factory.createDependencyArtifact(d), classpath, addDependencies);
     }
 
-    void addToClasspath(Artifact artifact, Set<String> classpath, boolean addDependencies) throws Exception {
+    void addToClasspath(Artifact artifact, Collection<String> classpath, boolean addDependencies) throws Exception {
```
Comment (davidB) [inline src/main/java/scala_maven/ScalaMojoSupport.java:518]: The Set is to avoid duplicate (and need to remove duplicate)

Comment (tombentley) [reply src/main/java/scala_maven/ScalaMojoSupport.java:518]: I'm happy to change the existing call sites use a `Set` if that's what you prefer, but the call sites currently use `List`. Classpath order (potentially) matters: If there are classes with the same name but different bytecode on the path then the order is relevant (I've not idea of the rest of maven really cares much about this TBH). So changing the call sites does introduce the possibility of subtle breakage for existing users. Your call.

Comment (davidB) [reply src/main/java/scala_maven/ScalaMojoSupport.java:518]: when the command send to scalac is sent, the plugin sort the classpath to have reproducible build (I opened the issue on scalac 2.7 .... I though it was solved). iirc the sort is done after the full classpath is defined (and before calling scalac....)

**Comment by tombentley** [comment]
I changed it to `Set`, initially only locally to the callers. But I made another commit to pass these classpaths as `Sets` more widely. That's a bit messy (e.g. `MavenProject.getCompileClasspathElements()` and `MavenProject.getTestClasspathElements()` return `Lists` so there's some unavoidable copying), but seems to fit better with what you're asking for. I can easily drop the latter commit if you prefer the former approach.

**Comment by tombentley** [comment]
Oh, @slandelle I made your changes too, if you want to take another look.

File: src/main/java/scala_maven/ScalaDocMojo.java [hunk 8]
```diff
@@ -1,10 +1,13 @@
 package scala_maven;
 
+import javax.print.attribute.HashAttributeSet;
```
Comment (slandelle) [inline src/main/java/scala_maven/ScalaDocMojo.java:3]: ?!

File: src/main/java/scala_maven/ScalaSourceMojoSupport.java [hunk 9]
```diff
@@ -130,4 +140,12 @@ private void initFilters() throws Exception {
         }
         return newCompileSourceRootsList;
     }
+
+    void addAdditionalDependencies(Set<String> back) throws Exception {
```
Comment (slandelle) [inline src/main/java/scala_maven/ScalaSourceMojoSupport.java:144]: protected

File: src/main/java/scala_maven/ScalaDocMojo.java (hunk 8 above)
Comment (tombentley) [reply src/main/java/scala_maven/ScalaDocMojo.java]: Damn autocomplete ;-)

**Comment by tombentley** [comment]
@slandelle those are now fixed, plus removed some unused imports added by a previous commit.

**Comment by davidB** [comment]
Thanks

**Comment by tombentley** [comment]
Thanks to both of you! Do you know when the next version of the plugin might be released?

**Comment by slandelle** [comment]
Release in progress (will be my first for this project, crossing fingers...)

**Comment by tombentley** [comment]
Awesome, thank you!

**Comment by slandelle** [comment]
4.4.0 is up on maven central.

**Comment by davidB** [comment]
@slandelle  Hourra! Your first release, I hope without too much trouble.
@tombentley Thanks for your contributions, next time you could add your name in the contributors list ;-)

**Comment by slandelle** [comment]
> I hope without too much trouble

It went smoothly. I'm just considering adding a configuration to skip the tests. Having to run them twice is sooooooo slow.

**Pull request merged**